### ISSUE TITLE
Server Crashes When Signing In

### ISSUE THREAD

**Issue by merktassel** [MKS-PostgreSQL/open2#18]
I consistently get this error when signing in with known credentials. I'm not sure if it's due to us changing something, as it worked fine during their demo. I'm going to try and test it after each commit to ensure it's not anything that we did.

```
Magic happens on 8080
Database is connected
This is our password in our db $2a$10$gtA2IGTFzLmQpAJqQ.

/open2/server/node_modules/mysql/lib/protocol/Parser.js:77
        throw err; // Rethrow non-MySQL errors
        ^
Not a valid BCrypt hash.
```

**Comment by merktassel** [comment]
I consistently get this error when signing in with known credentials. I'm not sure if it's due to us changing something, as it worked fine during their demo. I'm going to try and test it after each commit to ensure it's not anything that we did.

**Comment by merktassel** [comment]
I solved the issue listed above. For some reason [bcrypt](https://www.npmjs.com/package/bcrypt) was being used to hash the passwords when signing up, but [bcrypt-nodejs](https://www.npmjs.com/package/bcrypt-nodejs) was being used to compare the hashes when signing in. The two must have slightly different ways of hashing, so that's why `bcrypt-nodejs` kept complaining about the hash being invalid.

Apparently `bcrypt-nodejs` is just a fork of `bcrypt` that someone created because it the original didn't work very well on Windows. `bcrypt` is by far the more popular repository, and has much clearer documentation. Because of this, I switched both files over to `bcrypt`.

However, that led to another issue — the server would constantly say passwords were incorrect. After a lot of debugging, I found out the hashed password being retrieved from the database was different from what we inserted. The schema limits the password field to twenty five characters, which is significantly less than the size of the actual password. So you'd insert a fifty (or so) character hash, but it'd only store the first twenty five characters:

Insertion

```
username -> herpderp
password -> herpderp
hashed password -> $2a$10$h7ExZ3idEfwUrdGvArZGMOQCmmrbQI2C06j0oSTPSdFOY2QM15h/m
```

Retrieval

```
username -> herpderp
password -> herpderp
hashed password -> $2a$10$h7ExZ3idEfwUrdGvAr
```

Both of these issues have been fixed. The only thing left to do is clear the broken hashes from our tables.
